### DISCUSSION TITLE
(German) Train Data not in correct format

### DISCUSSION THREAD

**Discussion by ASB1993** [explosion/spaCy#7990] [Help: Coding & Implementations] [answered]
I really could need some help with creating training data for spacy. I tried many ways in creating training data for spacy. I started with a csv of words and entities, converted them to list of words and entities, putting the words together to lists of sentences and the tags to lists of tags per sentence. I then converted them to the json format. I now have multiple versions of json files that I wanted to convert to the new .spacy format. However, it seems as if no training data works after using --converter ner as it does not find NER format.

I first tried to convert the example to a json file

```
next_sentence = ""
word_index_in_sentence = 0
start_index = list()
end_index = list()
sent_tags = list()
TRAIN_DATA = []
with open("/content/drive/MyDrive/train_file.json", "w+", encoding="utf-8") as f:
    for word_index, word in enumerate(word_list):
        if word_index_in_sentence is 0:
            start_index.append(0)
        else:
            start_index.append((end_index[word_index_in_sentence-1])+1)

        sent_tags.append(tag_list[word_index])

        if word == "." or word == "?" or word == "!" or word_index == len(word_list)-1:
            next_sentence += word
            end_index.append(start_index[word_index_in_sentence]+1)
            entities = "";
            for i in range(word_index_in_sentence):
                if (i != 0):
                    entities += ","
                entities += "(" + str(start_index[i]) + "," + str(end_index[i]) + "," + "'" + sent_tags[i] + "'" + ")"

            f.write('("' + next_sentence + '",{"entities": [' + entities + ']}),')
            next_sentence = ""
            word_index_in_sentence = 0
            start_index = list()
            end_index = list()
            sent_tags = list()
        else:
            if word_list[word_index + 1] == "," or word_list[word_index + 1] == "." or word_list[word_index + 1] == "!" or word_list[word_index + 1] == "?":
                next_sentence += word
                end_index.append(start_index[word_index_in_sentence]+len(word)-1)
            else:
                next_sentence += word + " "
                end_index.append(start_index[word_index_in_sentence]+len(word))
            word_index_in_sentence += 1
Since this did not work as expected. I then tried to write a list of dicts of dicts. So instead of

f.write('("' + next_sentence + '",{"entities": [' + entities + ']}),')
I created a list TRAIN_DATA, adding the values as dict like this:

TRAIN_DATA.append({next_sentence: {"entities":entities}})
```
saving TRAIN_DATA again to a json-file.

However, when using `python -m spacy convert --converter ner /path/to/file  /path/to/save` it converts it to .spacy, nevertheless, it states:

⚠ Can't automatically detect NER format. Conversion may not succeed. See https://spacy.io/api/cli#convert ⚠ No sentence boundaries found to use with option -n 1. Use -s to automatically segment sentences or -n 0 to disable. ⚠ No sentence boundaries found. Use -s to automatically segment sentences. ⚠ No document delimiters found. Use -n to automatically group sentences into documents. ✔ Generated output file (1 documents): /content/drive/MyDrive/TRAIN_DATA/hope.spacy

My Training Data is either looking like this after converting to json:

> [{"Schartau sagte dem Tagesspiegel vom Freitag, Fischer sei in einer Weise aufgetreten, die alles andere als überzeugend war.": {"entities": "(0,8,'B-PER'),(9,14,'O'),(15,18,'O'),(19,31,'B-ORG'),(32,35,'O'),(36,42,'O'),(43,44,'O'),(45,52,'B-PER'),(53,56,'O'),(57,59,'O'),(60,65,'O'),(66,71,'O'),(72,82,'O'),(83,84,'O'),(85,88,'O'),(89,94,'O'),(95,101,'O'),(102,105,'O'),(106,117,'O'),(118,120,'O')"}}, {"welt.de vom 29.10.2005 Firmengründer Wolf Peter Bree arbeitete Anfang der siebziger Jahre als Möbelvertreter, als er einen fliegenden Händler aus dem Libanon traf.": {"entities": "(0,22,'[2005-10-29]'),...

or like this:

`[("Schartau sagte dem Tagesspiegel vom Freitag, Fischer sei in einer Weise aufgetreten, die alles andere als überzeugend war.", {"entities": (0,8,'B-PER'),(9,14,'O'),(15,18,'O'),(19,31,'B-ORG'),(32,35,'O'),(36,42,'O'),(43,44,'O'),(45,52,'B-PER'),(53,56,'O'),(57,59,'O'),(60,65,'O'),(66,71,'O'),(72,82,'O'),(83,84,'O'),(85,88,'O'),(89,94,'O'),(95,101,'O'),(102,105,'O'),(106,117,'O'),(118,120,'O')}),....
`
python -m spacy debug data /path/to/config
gives me the output:

⚠ The debug-data command is now available via the 'debug data' subcommand (without the hyphen). You can run python -m spacy debug --help for an overview of the other available debugging commands.

============================ Data file validation ============================ ✔ Corpus is loadable ✔ Pipeline can be initialized with data

=============================== Training stats =============================== Language: de Training pipeline: transformer, ner 1 training docs 1 evaluation docs ✔ No overlap between training and evaluation data ✘ Low number of examples to train a new pipeline (1)

============================== Vocab & Vectors ============================== ℹ 1 total word(s) in the data (1 unique) ℹ No word vectors present in the package

========================== Named Entity Recognition ========================== ℹ 1 label(s) 0 missing value(s) (tokens with '-' label) ⚠ Low number of examples for label 'stamt",{"entities":[(0,51,"O"),(52,67,"B' (1) ⚠ No examples for texts WITHOUT new label 'stamt",{"entities":[(0,51,"O"),(52,67,"B' ✔ No entities consisting of or starting/ending with whitespace ✔ No entities consisting of or starting/ending with punctuation

================================== Summary ================================== ✔ 5 checks passed ⚠ 2 warnings ✘ 1 error

Can someone PLEASE help me to convert my list of words and entities to spacys NER format to train a NER? I would appreciate it. Thank you!

- Operating System: Windows
- Python Version Used: 3.8
- spaCy Version Used: 3
- Environment Information: Google Colab

**Comment by svlandeg** [comment]
Hi! As a maintenance note: this is a good question for the discussions forum, so I'll move it there. The original issue will be closed and will link through to the open discussion thread.

**Comment by polm** [comment, marked as answer]
If you have data in spaCy v2 json format, you should use `--converter json`, not `--converter ner`, even if your data includes NER data.

I can see how this would be confusing, but "NER" in this case describes the format, not the contents of the data. "NER" here refers to a certain minimal format sometimes used specifically for NER annotations.

Debug data seems to be pointing out other issues with your data, but for starters using the `--converter json` command should let you at least get the format right.

**Reply by adrianeboyd** [reply]
To jump in: for the converter, JSON only refers to the v2 JSON training format (https://v2.spacy.io/api/annotation#json-input), not the `TRAIN_DATA`-ish format as JSON. The v2 JSON training format only supports IOB tags for entities, not character offsets.

I would suggest having a look at the `convert` step in the demo project `pipelines/ner_demo`:

Project:

https://github.com/explosion/projects/tree/v3/pipelines/ner_demo

Demo data:

https://github.com/explosion/projects/tree/v3/pipelines/ner_demo/assets

Demo convert script:

https://github.com/explosion/projects/blob/v3/pipelines/ner_demo/scripts/convert.py

The input data is the `TRAIN_DATA` format saved as JSON and you can see how to convert that format to `Doc` and `DocBin`.

**Reply by ASB1993** [reply]
> If you have data in spaCy v2 json format, you should use `--converter json`, not `--converter ner`, even if your data includes NER data.

Thanks for the input. However, I am trying to convert my data, which is not working well. I tried the Doc and DocBin approach, however, my data is not readable, so if I have string-entities, how could I convert them from them being still in the correct format?

**Reply by adrianeboyd** [reply]
Ah, I think I see where some of the confusion might be. There are a couple things going wrong.

First, make sure that `"entities"` is a list of tuples or a list IOB tags, not a string that looks like a list of tuples.

Second, if you're using character offsets, then you provide the whole entity span, not IOB tags, and you only include entities, not `O`:

Character offsets look like this and refer to the raw text:

```python
"entities": [(0, 8, "PER"), (45, 52, "PER"), (89, 105, "MISC")]
```

See examples here of both the character offset version (what I think is what you should use here) and also the IOB version. Scroll down a bit to a black examples block in the main column: https://spacy.io/api/data-formats#dict-input

If you can include a single python conversion script with the data for one text+annotation that we can run, that would be the easiest to debug.

**Reply by ASB1993** [reply]
[NER-de-train.zip](https://github.com/explosion/spaCy/files/6423054/NER-de-train.zip)
[file_json.zip](https://github.com/explosion/spaCy/files/6423367/file_json.zip)
[train_file.txt](https://github.com/explosion/spaCy/files/6423374/train_file.txt)


So here is my train data in IOB format. However, using the converter is not working either, as it states "ValueError: [E177] Ill-formed IOB input detected: 30". I tried to take a look and I deleted some rows, however, I have no idea what is wrong with the data. That is, why I tried to convert the file manually to the json format. 

I don't really have a script there, since I am using Google Colab. 
```
"""turn words to sentence train file"""

import csv
import sys
import json
import pandas as pd
import numpy as np
import spacy
import re



csvFilePath = "/content/drive/MyDrive/germaner/NER-de-train.tsv"
jsonFilePath = "/content/drive/MyDrive/germaner/NER-de-train.txt"


df = pd.read_csv(csvFilePath, sep='\t', names=["words","tags"])


word_list = list(df["words"])
tag_list = list(df["tags"])
"""transform train data to spacy train data"""
next_sentence = ""
word_index_in_sentence = 0
start_index = list()
end_index = list()
sent_tags = list()
sentences = []
tags = []
TRAIN_DATA = []
sentence_count = []
n = 0
with open("/content/drive/MyDrive/train_file.json", "w+", encoding="utf-8") as f:
    for word_index, word in enumerate(word_list):
        if word_index_in_sentence is 0:
            start_index.append(0)
        else:
            start_index.append((end_index[word_index_in_sentence-1])+1)

        sent_tags.append(tag_list[word_index])

        if word == "." or word == "?" or word == "!" or word_index == len(word_list)-1:
            next_sentence += word
            end_index.append(start_index[word_index_in_sentence]+1)
            entities = "";
            for i in range(word_index_in_sentence):
                if (i != 0):
                    entities += ","
                entities += "(" + str(start_index[i]) + "," + str(end_index[i]) + "," + "'" + sent_tags[i] + "'" + ")"
                
            # TRAIN_DATA.#
            n += 1
            sentences.append(next_sentence)
            sentence_count.append(n)
            tags.append(entities)

            f.write('("' + next_sentence + '",{"entities": [' + entities + ']}),')
            next_sentence = ""
            word_index_in_sentence = 0
            start_index = list()
            end_index = list()
            sent_tags = list()
        else:
            if word_list[word_index + 1] == "," or word_list[word_index + 1] == "." or word_list[word_index + 1] == "!" or word_list[word_index + 1] == "?":
                next_sentence += word
                end_index.append(start_index[word_index_in_sentence]+len(word)-1)
            else:
                next_sentence += word + " "
                end_index.append(start_index[word_index_in_sentence]+len(word))
            word_index_in_sentence += 1
```

I also tried this:

```
import srsly
import typer
import warnings
from pathlib import Path
from spacy.tokens import DocBin

f = open("/content/drive/MyDrive/train_file.json")
data = json.loads(f)  # also tried json.loads(f.read())

nlp = spacy.blank("de")
db = DocBin()

for text, annot in srsly.read_json("/content/drive/MyDrive/train_file.json"):
    doc = nlp.make_doc(text)
    ents = []
    for start, end, label in annot["entities"]:
        span = doc.char_span(start, end, label=label)
        if span is None:
            msg = f"skip entity"
            warnings.warn(msg)
        else:
            ents.append(span)
    doc.ents = ents
    db.add(doc)
db.to_disk("/content/drive/MyDrive/different_train.spacy")

```

nevertheless, it does not open the json-file. 

I have no idea, how I should provide a conversion file? I am pretty new at writing issues.

Thank you for your help!!

**Reply by polm** [reply]
Ah, well now it's clear what your problem is. In your tsv data you have sentences separated by blank lines, which is OK, but at the start of each sentence (except the first one?) you have a line that isn't part of the sentence - it looks like the sentence source. Example:

    welt.de vom 29.10.2005  [2005-10-29]

spaCy can't parse that. If you get rid of those lines the NER converter should just work.

**Reply by ASB1993** [reply]
[Repaired-NER-de-train.zip](https://github.com/explosion/spaCy/files/6426396/Repaired-NER-de-train.zip)

Hi, thanks for your advice.
I now got rid of those lines. But it still does not work as it gives me the same error message. I included the new file. What is now wrong with my data?

**Reply by adrianeboyd** [reply]
Here's the original format (4 tab-separated columns with comment lines starting with `#`):

```none
#	n-tv.de vom 26.02.2005	[2005-02-26]	
1	Schartau	B-PER	O
2	sagte	O	O
3	dem	O	O
4	"	O	O
5	Tagesspiegel	B-ORG	O
6	"	O	O
7	vom	O	O
8	Freitag	O	O
9	,	O	O
10	Fischer	B-PER	O
11	sei	O	O
12	"	O	O
13	in	O	O
14	einer	O	O
15	Weise	O	O
16	aufgetreten	O	O
17	,	O	O
18	die	O	O
19	alles	O	O
20	andere	O	O
21	als	O	O
22	überzeugend	O	O
23	war	O	O
24	"	O	O
25	.	O	O
```

You want to:

* skip lines that start with `#`
* extract columns 2 and 3 without any changes to formatting/encoding and output the two columns with tab separation like this (no extra quotes, everything as UTF-8):
    ```none
    word	O
    ```
* keep blank lines between sentences

Then you can convert the resulting output with `spacy convert -c ner`.

**Reply by ASB1993** [reply]
Thank you Adriane.

I just got it, thank you so much for your help!!! AWESOME!!!

**Comment by mbrunecky** [comment]
In Spacy2 discussions, it has been suggested that Spacy 'convert' utility should support conversion from the '**TRAIN_DATA**' format, which is very appealing to NER people:
`TRAIN_DATA = ["Who is Shaka Khan?", {"entities": [(7, 17, "PERSON")]})]`

Generating the other 'convert' utility input data formats is more involved, and (in my opinion) demands that you replicate Spacy tokenization - without it you can not always break your text into tagged tokens correctly.

I  have been using Spacy (with srsly) to read the TRAIN_DATA format from JSONL files (one or more documents per file), and in Spacy2 converting them to Spacy2 JSON format. Now in Spacy3 I added converting them to binary DocBin data (along with batching multiple documents into one DocBin). I can also generate the raw_text format used for pre-training.

The following are the essential code fragments from my conversion utility (with due apologies - Python is not my language yet):

```
    nlp = spacy.blank("en")
    docbin = DocBin()

    for data in srsly.read_jsonl(train_data_file):
        text  = data[0]
        annots = data[1]
        if annots.get('entities') == None:
            continue
        doc  = nlp(text)
        tags = offsets_to_biluo_tags(doc, annots['entities'])
        doc.ents = biluo_tags_to_spans(doc, tags)
        docbin.add(doc)
        # add batching multiple documents into a DocBin here
        docbin.to_disk(db_file)
        docbin = DocBin()

```
The advantage of using nlp = spacy.blank("en") and  doc=nlp(text) is that it will do the tokenization for you. Unless you have very well-delimited data, you may find that assuring your start/end offsets fall on token start/ends may require some extra effort. My OCR data is not always well-delimited, and I must insert delimiting where it 'does not fit'. In such case, the  offsets_to_biluo_tags() reports an error. Unfortunately, the error report is not very helpful:
```
ERROR in 'G:\Data\ML\ca_placer_dee_10000_sel_7\model/trn_dee_6284310.jsonl': [W030] Some entities could not be aligned in the text "RECORDING REQUESTED BY:
First American Title Compa..." with entities "[[1095, 1111, 'NAME_FROM'], [1116, 1134, 'NAME_FRO...". Use `spacy.training.offsets_to_biluo_tags(nlp.make_doc(text), entities)` to check the alignment. Misaligned entities ('-') will be ignored during training.
```

Rather than reporting the document text and ALL annotations, it would help reporting the '**wrong**' annotation.

**Reply by adrianeboyd** [reply]
The original dataset here doesn't have character offsets, so I think it's easiest to convert it to the CoNLL-2003 tab-separated format and use `spacy convert`. I guess it does depend a bit on whether you want to try to do the detokenization to try to create realistic raw texts rather than working with pretokenized text, though. `spacy` does support training from texts without known whitespace between the tokens, but I don't think the v3 NER converters actually set the right feature for this. (The `ner` and `iob` converters should probably set `doc.has_unknown_spaces`.)

`spacy convert` has never supported the `TRAIN_DATA` format from the v2 demo scripts. In v2 it supported a JSONL format that came from prodigy that had similar spans with character offsets.

Here's a demo script that converts from the `TRAIN_DATA` format saved as JSON:

https://github.com/explosion/projects/blob/v3/pipelines/ner_demo/scripts/convert.py

It gives errors for the individual entities that can't be aligned. You can see it run in the context of the whole demo project [`pipelines/ner_demo`](https://github.com/explosion/projects/blob/v3/pipelines/ner_demo/scripts/convert.py).
